src/mscorlib/src/System/Lazy.cs

```diff
@@ -202,29 +239,171 @@ public Lazy(Func<T> valueFactory, bool isThreadSafe)
+                    throw;
+                }
+            }
+            return m_value;
         }
 
+        internal T PopulateValueExecutionAndPublication(LazyImplementation sync, bool useDefaultConstructor)
+        {
+            lock (sync) // we're safe to lock on "this" as object is an private object used by Lazy
+            {
+                // it's possible for multiple calls to have piled up behind the lock, so we need to check
+                // to see if the ExecutionAndPublication object is still the current implementation.
+                if (ReferenceEquals(m_implementation, sync))
+                    return PopulateValue(LazyThreadSafetyMode.ExecutionAndPublication, useDefaultConstructor);
+                return Value;
+            }
+        }
+
+        /// <summary>
+        /// Creates the underlying value using the factory object into a helper object and, for the
+        /// first object to complete its factory, uses that objects LazyWorker&lt;T&gt; implementation
+        /// </summary>
+        /// <param name="factory">The object factory used to create the underlying object</param>
+        /// <param name="comparand">The publication object, used for synchronisation and comparison</param>
+        /// <returns>The underlying object</returns>
+        internal T PopulateValuePublicationOnly(LazyImplementation initializer, LazyImplementation spinner, bool useDefaultConstructor)
+        {
+            T possibleValue;
+            if (useDefaultConstructor)
+            {
+                possibleValue = CreateViaDefaultConstructor();
+            }
+            else
+            {
+                var factory = m_factory;
+                if (factory == null)
+                    return spinner.GetValue(this);
+                possibleValue = factory();
+            }
+
+            try { }
+            finally
+            {
+                // we run this in a finally block to ensure that we don't get a partial completion due
+                // to a Thread.Abort, which could mean that other threads might be left in infinite loops
+                var previous = Interlocked.CompareExchange(ref m_implementation, spinner, initializer);
+                if (previous == initializer)
+                {
+                    m_value = possibleValue;
+                    m_factory = null;
+                    m_implementation = null;
+                }
+            }
+
+            return Value;
+        }
+
+        internal void WaitForPublicationOnly()
+        {
+            while (!ReferenceEquals(m_implementation, null))
+            {
+                // CreateValuePublicationOnly temporarily sets m_implementation to "this". The Lazy implementation
+                // has an explicit iplementation of LazyWorker which just waits for the m_value to be set, which is
+                // signalled by m_implemenation then being set to null.
+                Thread.Sleep(0);
```

Comment:
`SpinWait` would be better - it handles the different corner cases for which `Thread.Sleep(0)` does not work well.